src/ucp/core/ucp_ep.c

```diff
@@ -1143,10 +1144,38 @@ static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
     }
 }
 
+static int ucp_ep_can_unprogress_lane(ucp_ep_h ep, uct_ep_h uct_ep)
+{
+    if (!ep->worker->context->config.ext.adaptive_progress) {
+        return 0;
+    }
+
+    if (!ucp_ep_has_cm_lane(ep)) {
+        return 0;
+    }
+
+    if (ucp_is_uct_ep_failed(uct_ep)) {
+        return 0;
+    }
+
+    if (!ucp_wireup_ep_test(uct_ep)) {
+        return 1;
+    }
```

Comment:
I think all simple negative conditions could be united to single ORed if-statement

Reply:
done